From 4555ba7c21735333544dc484f3eac14a7cb61b51 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Thu, 31 Jul 2025 18:51:42 +0200
Subject: [PATCH] Detach dialogs from the desktop

Set dialogs as modal with normal window type, and add them to individual
window groups. This ensures that dialogs are kept visible on Wayland, but
not lock or attached to the desktop. Also, set prgname to
'org.gnome.Shell.Extensions' so dialogs will have proper icon on the
taskbar/window switcher.
---
 app/autoAr.js         |  9 +++++++--
 app/desktopManager.js |  1 -
 app/ding.js           |  2 ++
 app/fileItemMenu.js   | 11 +++++++----
 app/preferences.js    |  8 +++++---
 5 files changed, 21 insertions(+), 10 deletions(-)

diff --git a/app/autoAr.js b/app/autoAr.js
index cac7143c..cb4b86f3 100644
--- a/app/autoAr.js
+++ b/app/autoAr.js
@@ -17,6 +17,7 @@
 'use strict';
 const GLib = imports.gi.GLib;
 const Gio = imports.gi.Gio;
+const Gdk = imports.gi.Gdk;
 const Gtk = imports.gi.Gtk;
 var GnomeAutoar = null;
 try {
@@ -40,10 +41,12 @@ var AutoAr = class {
             title: 'Archives Operations',
             resizable: false,
             deletable: false,
-            modal: false,
+            modal: true,
             default_height: 100,
             window_position: Gtk.WindowPosition.CENTER_ALWAYS,
         });
+        const windowGroup = new Gtk.WindowGroup();
+        windowGroup.add_window(this._progressWindow);
         this._progressWindow.connect('delete-event', () => {
             return true;
         });
@@ -552,14 +555,16 @@ const CompressDialog = class {
         this._destinationFolder = destinationFolder;
         this._dialog = new Gtk.Dialog({
             title: _('Create archive'),
-            transientFor: parentWindow,
             resizable: false,
             modal: true,
             use_header_bar: true,
             default_width: 500,
             default_height: 210,
             window_position: Gtk.WindowPosition.CENTER_ALWAYS,
+            type_hint: Gdk.WindowTypeHint.NORMAL,
         });
+        const windowGroup = new Gtk.WindowGroup();
+        windowGroup.add_window(this._dialog);
         const container = this._dialog.get_content_area();
         container.orientation = Gtk.Orientation.VERTICAL;
         container.margin_top = 30;
diff --git a/app/desktopManager.js b/app/desktopManager.js
index b7829f0a..44097c4d 100644
--- a/app/desktopManager.js
+++ b/app/desktopManager.js
@@ -1097,7 +1097,6 @@ var DesktopManager = class {
     }
     findFiles(window, text) {
         this._findFileWindow = new Gtk.Dialog({
-            transientFor: window,
             use_header_bar: true,
             window_position: Gtk.WindowPosition.CENTER_ON_PARENT,
             resizable: false,
diff --git a/app/ding.js b/app/ding.js
index 9a1550ad..3bbcf941 100755
--- a/app/ding.js
+++ b/app/ding.js
@@ -159,6 +159,8 @@ if (Gio.File.new_for_path(localePath).query_exists(null)) {
     Gettext.bindtextdomain('ding', localePath);
 }
 
+GLib.set_prgname('org.gnome.Shell.Extensions');
+
 const DesktopManager = imports.desktopManager;
 
 var desktopManager = null;
diff --git a/app/fileItemMenu.js b/app/fileItemMenu.js
index fb11705c..93791d7d 100644
--- a/app/fileItemMenu.js
+++ b/app/fileItemMenu.js
@@ -412,10 +412,12 @@ var FileItemMenu = class {
             if (fileItems[0].isDirectory) {
                 mimetype = 'inode/directory';
             }
-            const chooser = Gtk.AppChooserDialog.new_for_content_type(
-                this._currentFileItem.container.get_toplevel(),
+            const chooser = Gtk.AppChooserDialog.new_for_content_type(null,
                 Gtk.DialogFlags.MODAL | Gtk.DialogFlags.USE_HEADER_BAR,
                 mimetype);
+            chooser.set_type_hint(Gdk.WindowTypeHint.NORMAL);
+            const windowGroup = new Gtk.WindowGroup();
+            windowGroup.add_window(chooser);
             let signals = new SignalManager.SignalManager();
             chooser.show_all();
             signals.connectSignal(chooser, 'close', () => {
@@ -468,14 +470,15 @@ var FileItemMenu = class {
         const dialog = new Gtk.FileChooserDialog({
             title: _('Select Extract Destination'),
             modal: true,
-            transientFor: this._currentFileItem.container.get_toplevel(),
+            type_hint: Gdk.WindowTypeHint.NORMAL,
         });
+        const windowGroup = new Gtk.WindowGroup();
+        windowGroup.add_window(dialog);
         dialog.set_action(Gtk.FileChooserAction.SELECT_FOLDER);
         dialog.set_create_folders(true);
         dialog.set_current_folder_uri(DesktopIconsUtil.getDesktopDir().get_uri());
         dialog.add_button(_('Cancel'), Gtk.ResponseType.CANCEL);
         dialog.add_button(_('Select'), Gtk.ResponseType.ACCEPT);
-        DesktopIconsUtil.windowHidePagerTaskbarModal(dialog, true);
         dialog.show_all();
         dialog.connect('close', () => {
             dialog.response(Gtk.ResponseType.CANCEL);
diff --git a/app/preferences.js b/app/preferences.js
index b25e1f2a..9e0498c4 100644
--- a/app/preferences.js
+++ b/app/preferences.js
@@ -87,17 +87,19 @@ function init(path) {
  */
 function showPreferences() {
     if (prefsWindow) {
+        prefsWindow.present();
         return;
     }
     prefsWindow = new Gtk.Window({
-        resizable: false,
+        title: _('DING Settings'),
+        modal: true,
         window_position: Gtk.WindowPosition.CENTER,
     });
     prefsWindow.connect('destroy', () => {
         prefsWindow = null;
     });
-    prefsWindow.set_title(_('Settings'));
-    DesktopIconsUtil.windowHidePagerTaskbarModal(prefsWindow, true);
+    const windowGroup = new Gtk.WindowGroup();
+    windowGroup.add_window(prefsWindow);
     let frame = PrefsWindow.preferencesFrame(Gtk, desktopSettings, nautilusSettings, gtkSettings);
     prefsWindow.add(frame);
     prefsWindow.show_all();
-- 
GitLab

